chainerx_cc/chainerx/cuda/cuda_device/batch_norm.cc

```diff
@@ -100,79 +103,90 @@ class CudnnBNTensorDescriptor {
     cudnnTensorDescriptor_t desc_{};
 };
 
-class CudaBatchNormForwardBackward : public chainerx::GenericBatchNormForwardBackward {
+struct CudaBatchNormState {
+    CudaBatchNormState(Array x_cont, Array x_mean, Array x_inv_std)
+        : x_cont{std::move(x_cont)}, x_mean{std::move(x_mean)}, x_inv_std{std::move(x_inv_std)} {}
+
+    Array x_cont;
+    Array x_mean;
+    Array x_inv_std;
+};
+
+}  // namespace
+
+class CudaBatchNormForwardOp : public BatchNormForwardOp {
 public:
-    explicit CudaBatchNormForwardBackward(
-            cuda_internal::CudnnHandle& cudnn_handle,
+    void Call(
+            const Array& x,
+            const Array& gamma,
+            const Array& beta,
             const Array& running_mean,
             const Array& running_var,
             Scalar eps,
             Scalar decay,
-            const Axes& axis)
-        : GenericBatchNormForwardBackward{running_mean, running_var, eps, decay, axis}, cudnn_handle_{cudnn_handle} {
-        if (static_cast<double>(eps) < CUDNN_BN_MIN_EPSILON) {
-            throw CudnnError{"Minimum allowed epsilon is ", CUDNN_BN_MIN_EPSILON, " but found ", eps, "."};
-        }
-        if (!running_mean.IsContiguous()) {
-            throw DeviceError{"Running mean must to be contiguous for cuDNN to update it in-place."};
-        }
-        if (!running_var.IsContiguous()) {
-            throw DeviceError{"Running variance must to be contiguous for cuDNN to update it in-place."};
-        }
-    }
-
-    Array Forward(const Array& x, const Array& gamma, const Array& beta) override {
+            const Axes& axis,
+            const Array& out,
+            nonstd::optional<std::shared_ptr<void>>& state) override {
```

Comment:
Thanks for the suggestion. How about using `std::shared_ptr<BatchNormState>` and keep it as an argument? That would allow the caller to specify whether the state is needed or not (pass `nullptr` in case caller don't care) without any extra arguments.

Reply:
How about passing a boolean flag (like `need_state`)?

Reply:
Changed to `std::shared_ptr<BatchNormState>` with a boolean argument `return_state`.